modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -156,57 +156,99 @@ public IgniteRel visit(IgniteTableScan rel) {
-            }
+            prevProjects.add(rel.getProjects());
 
             return super.visit(rel);
         }
 
         /** {@inheritDoc} */
         @Override
         public IgniteRel visit(IgniteValues rel) {
-            if (expressions == null) {
-                expressions = new ArrayList<>();
-            }
+            List<List<RexNode>> expressions = Commons.cast(rel.getTuples());
+
+            if (!prevProjects.isEmpty()) {
+                for (List<RexNode> prj : prevProjects) {
+                    List<RexNode> prevProject = prj;
+                    List<RexNode> projectionsReplaced = replaceInputRefs(prevProject);
+
+                    boolean refFound = !projectionsReplaced.equals(prevProject);
+
+                    prevProject = projectionsReplaced;
+
+                    assert rel.getTuples() != null;
 
-            List<List<RexNode>> values = Commons.cast(rel.getTuples());
+                    for (List<RexNode> exp : expressions) {
+                        if (!refFound) {
+                            // no references are found, all important in projections
+                            finalExpressions.add(prevProject);
+                        } else {
+                            // otherwise exchange references with appropriate representations
+                            List<RexNode> modifiedExpressions = new ArrayList<>(prevProject);
 
-            expressions.addAll(values);
+                            for (int i = 0; i < prevProject.size(); ++i) {
+                                RexNode prjNode = prevProject.get(i);
+
+                                if (prjNode instanceof RexLocalRef) {
+                                    RexLocalRef node0 = (RexLocalRef) prjNode;
+                                    modifiedExpressions.set(i, exp.get(node0.getIndex()));
+                                }
+                            }
+
+                            finalExpressions.add(modifiedExpressions);
+                        }
+                    }
+                }
+            } else if (expressions != null) {
+                finalExpressions.addAll(expressions);
+            }
+
+            prevProjects.clear();
 
             return super.visit(rel);
         }
 
         /** {@inheritDoc} */
         @Override
         public IgniteRel visit(IgniteUnionAll rel) {
-            unionRaised = true;
+            if (mapping != null) {
+                // unexpected multiple unions
+                throw Util.FoundOne.NULL;
+            }
+
+            for (List<RexNode> prj : prevProjects) {
+                if (mapping == null) {
+                    mapping = new ArrayList<>(prevProjects.size());
+                }
+
+                mapping.add(RexUtils.inversePermutation(prj,
+                        table.getRowType(Commons.typeFactory()), false));
+            }
+
+            prevProjects.clear();
+
             return super.visit(rel);
         }
 
```

Comment:
Why in method `visitChild` we are checking  `previous` instead of `parent`? 

```
        protected void visitChild(IgniteRel parent, int i, IgniteRel child) {
            if (child instanceof IgniteExchange || child instanceof IgniteTrimExchange) {
                super.visitChild(parent, i, child);
            } else if (allowRelTransfers.getOrDefault(child.getClass(), Set.of()).contains(previous.getClass())) {
                super.visitChild(parent, i, child);
            } else {
                throw Util.FoundOne.NULL;
            }
        }
 ```

Reply:
fix it